DS13/code/modules/overmap/warpcore.dm

```diff
@@ -0,0 +1,676 @@
+	var/area/A = get_area(src)
+	var/obj/structure/overmap/F
+	if(A.linked_overmap)
+		F = A.linked_overmap //relay some sounds to the crew.
+	if(F)
+		F.warp_core = src
+		linked = F
+
+/datum/looping_sound/warp
+	start_length = 0
+	mid_sounds = list('DS13/sound/effects/warp_hum.ogg')
+	mid_length = 20
+	volume = 100
+
+/obj/machinery/power/warp_core/update_icon()
+	if(!dilithium_matrix)
+		icon_state = "core-empty"
+		return
+	if(!active)
+		icon_state = "core"
+		return
+	else
+		icon_state = "core-on"
+	if(containment < 0)
+		icon_state = "core-fast"
+	switch(containment)
+		if(0 to 19)
+			icon_state = "core-fast"
+		if(20 to 50)
+			icon_state = "core-medium"
+
+/obj/machinery/power/warp_core/attackby(obj/item/I, mob/user)
+	if(istype(I, /obj/item/dilithium_matrix) && !dilithium_matrix)
+		var/obj/item/dilithium_matrix/M = I
+		if(!M.stored)
+			to_chat(user, "<span class='warning'>That dilithium matrix lacks a dilithium crystal!</span>")
+			return ..()
+		to_chat(user, "<span class='notice'>You start to slot [I] into [src]</span>")
+		if(do_after(user, 50, target=src))
+			playsound(src, 'DS13/sound/effects/FTL/load.ogg', 50, 1)
+			I.forceMove(src)
+			dilithium_matrix = I
+			update_icon()
+			start()
+			return FALSE
+	. = ..()
+
+/obj/machinery/power/warp_core/Destroy()
+	QDEL_NULL(soundloop)
+	QDEL_NULL(Radio)
+	. = ..()
+
+/obj/machinery/power/warp_core/proc/start()
+	if(!active())
+		return
+	active = TRUE
+	soundloop.start()
+
+/obj/machinery/power/warp_core/proc/stop()
+	soundloop.stop()
+	active = FALSE
+
+/obj/machinery/power/warp_core/proc/active()
+	if(QDELETED(MI) || !MI)
+		MI = null
+		return FALSE
+	if(QDELETED(AI) || !AI)
+		AI = null
+		return FALSE
+	if(!matter_stream || !antimatter_stream)
+		return FALSE
+	return TRUE
+
+/obj/machinery/power/warp_core/process()
+	update_icon()
+	if(!linked)
+		find_overmap()
+	if(!active) //Not turned on. Don't process.
+		return
+	if(!outlet)
+		outlet = locate(/obj/machinery/atmospherics/components/binary/pump) in get_step(src, SOUTH)
+	if(!outlet.airs || !active())
+		stop()
+		return
+	check_failure()
+	announce_radio()
+	if(!dilithium_matrix)
+		return
+	var/penalty = 0.1 //You need to realign every 30 minutes or so.
+	dilithium_matrix.absorb_damage(penalty)
+	for(var/datum/gas_mixture/S in outlet.airs)
+		if(S.gases[/datum/gas/plasma])
+			S.gases[/datum/gas/plasma][MOLES] += 10 //Add some extra plasma to it.
+			S.garbage_collect()
+	var/power = 500000
+	add_avail(power)
+
+/obj/machinery/power/warp_core/proc/reset_voice()
+	voice_ready = TRUE
+
+/obj/machinery/power/warp_core/proc/announce_radio() //Warn the engis of specific events.
+	if(!voice_ready)
+		return FALSE
+	switch(containment)
+		if(0 to 20) //When it hits 20%. The engis need to know it's almost the point of no return.
+			Radio.talk_into(src,"WARNING: Antimatter containment failure imminent! Evacuate engineering immediately!. Containment is at [containment]%.","Engineering",get_spans(),get_default_language())
```

Comment:
These 4 lines are repeated for each case of the switch. They can be refactored to a procedure.